src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -113,6 +119,193 @@ static void ListConcatUniqueAttributeClassMemberLists(AttributeEquivalenceClass
+ * Note that the function expects (and asserts) the input query to be a top
+ * level union query defined by TopLevelUnionQuery().
+ */
+bool
+SafeToPushdownTopLevelUnion(RelationRestrictionContext *restrictionContext)
+{
+	Index unionQueryPartitionKeyIndex = 0;
+	AttributeEquivalenceClass *attributeEquivalance =
+		palloc0(sizeof(AttributeEquivalenceClass));
+	ListCell *relationRestrictionCell = NULL;
+
+	attributeEquivalance->equivalenceId = attributeEquivalenceId++;
+
+	foreach(relationRestrictionCell, restrictionContext->relationRestrictionList)
+	{
+		RelationRestriction *relationRestriction = lfirst(relationRestrictionCell);
+		Index partitionKeyIndex = InvalidAttrNumber;
+		PlannerInfo *relationPlannerRoot = relationRestriction->plannerInfo;
+		List *targetList = relationPlannerRoot->parse->targetList;
+		List *appendRelList = relationPlannerRoot->append_rel_list;
+		Var *varToBeAdded = NULL;
+		TargetEntry *targetEntryToAdd = NULL;
+
+		/* we first check whether UNION ALLs are pulled up */
+		if (appendRelList != NULL)
+		{
+			varToBeAdded = FindTranslatedVar(appendRelList,
+											 relationRestriction->relationId,
+											 relationRestriction->index,
+											 &partitionKeyIndex);
+
+			/* union does not have partition key in the target list */
+			if (partitionKeyIndex == 0)
+			{
+				return false;
+			}
+		}
+		else
+		{
+			partitionKeyIndex =
+				RelationRestrictionPartitionKeyIndex(relationRestriction);
+
+			/* union does not have partition key in the target list */
+			if (partitionKeyIndex == 0)
+			{
+				return false;
+			}
+
+			targetEntryToAdd = list_nth(targetList, partitionKeyIndex - 1);
+			if (!IsA(targetEntryToAdd->expr, Var))
+			{
+				return false;
+			}
+
+			varToBeAdded = (Var *) targetEntryToAdd->expr;
+		}
+
+		/*
+		 * If the first relation doesn't have partition key on the target
+		 * list of the query that the relation in, simply not allow to push down
+		 * the query.
+		 */
+		if (partitionKeyIndex == InvalidAttrNumber)
+		{
+			return false;
+		}
+
+		/*
+		 * We find the first relations partition key index in the target list. Later,
+		 * we check whether all the relations have partition keys in the
+		 * same position.
+		 */
+		if (unionQueryPartitionKeyIndex == InvalidAttrNumber)
+		{
+			unionQueryPartitionKeyIndex = partitionKeyIndex;
+		}
+		else if (unionQueryPartitionKeyIndex != partitionKeyIndex)
+		{
+			return false;
+		}
+
+		AddToAttributeEquivalenceClass(&attributeEquivalance, relationPlannerRoot,
+									   varToBeAdded);
+	}
+
+	return EquivalenceListContainsRelationsEquality(list_make1(attributeEquivalance),
+													restrictionContext);
+}
+
+
+/*
+ * FindTranslatedVar iterates on the appendRelList and tries to find a translated
+ * child var identified by the relation id and the relation rte index. The function
+ * returns NULL if it cannot find a translated var.
+ */
+static Var *
+FindTranslatedVar(List *appendRelList, Oid relationOid, Index relationRteIndex,
```

Comment:
Is it the Var was already translated, and we are trying to find translation ? 

Function could be better named. Could you rephrase explanation.